Source/Dafny/Resolver.cs

```diff
@@ -1585,7 +1585,41 @@ public class ModuleBindings
           if (useImports || string.Equals(kv.Key, "_default", StringComparison.InvariantCulture)) {
             TopLevelDecl d;
             if (sig.TopLevels.TryGetValue(kv.Key, out d)) {
-              sig.TopLevels[kv.Key] = AmbiguousTopLevelDecl.Create(moduleDef, d, kv.Value);
+              // ignore the import if the existing declaration belongs to the current module
+              if (d.Module != moduleDef) {
+                bool ok = false;
+                // keep just one if they normalize to the same type
+                if (d == kv.Value) {
+                  ok = true;
+                } else {
+                  TopLevelDecl dd = d;
+                  Type ddd = null;
+                  TopLevelDecl kk = kv.Value;
+                  Type kkk = null;
+                  if (d is TypeSynonymDecl && !(d is SubsetTypeDecl)) {
+                    ddd = (d as TypeSynonymDecl).Rhs.NormalizeExpand(true);
+                    if (ddd is UserDefinedType) {
+                      dd = (ddd as UserDefinedType).AsRevealableType as TopLevelDecl;
+                      ddd = null;
+                    }
+                  }
+                  if (kk is TypeSynonymDecl && !(kk is SubsetTypeDecl)) {
+                    kkk = (kk as TypeSynonymDecl).Rhs.NormalizeExpand(true);
+                    if (kkk is UserDefinedType) {
+                      kk = (kkk as UserDefinedType).AsRevealableType as TopLevelDecl;
+                      kkk = null;
+                    }
+                  }
+                  if (ddd == null || kkk == null) {
+                    ok = dd == kk;
+                  } else {
+                    ok = ddd.Equals(kkk, true);
+                  }
+                }
```

Comment:
I don't feel certain about this logic. It may be right, but I'm not convinced, so I'll at least ask some question here. One of the things that worries me is that it bypasses the mechanism for expanding types through module boundaries, where some things may be visible through imports and other things not.

I think the aim of this code is to repeatedly follow type synonyms to their non-synonym base definitions. That can be done with one call to `.NormalizeExpand(true)`, provided one starts with a `Type`, not a `TopLevelDecl`. So, then, can the code above instead be accomplished by:

```
var dType = UserDefinedType.FromTopLevelDecl(d, d.tok).NormalizeExpand(true);
var vType = UserDefinedType.FromTopLevelDecl(kv.Value, kv.Value.tok).NormalizeExpand(true);
var ok = dType.Equals(vType);
```

Or, for that matter, why not just

```
var dType = UserDefinedType.FromTopLevelDecl(d, d.tok);
var vType = UserDefinedType.FromTopLevelDecl(kv.Value, kv.Value.tok);
var ok = dType.Equals(vType, true);
```

Reply:
Used the last suggestion, with a bit of adjustment to accommodate the current use of .ResolvedParams for OpaqueTypeDecls.